File: cpp/next_greater_element.cpp
#include <bits/stdc++.h>

using namespace std;

/*
 * C++ stack, pop()出栈, top()返回栈顶元素, push(x) 入栈
 */
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> stk;
        unordered_map<int, int> d;
        reverse(nums2.begin(), nums2.end());
        for (int x : nums2) {
            while (!stk.empty() && stk.top() < x) {
                stk.pop();
            }
            if (!stk.empty()) {
                d[x] = stk.top();
            }
            stk.push(x);
        }
        vector<int> ans;
        for (int x : nums1) {
            ans.push_back(d.count(x) ? d[x] : -1);
        }
        return ans;
    }
};

int main() {
    Solution s;
    vector<int> nums1{4, 1, 2};
    vector<int> nums2{1, 3, 4, 2};
    auto ans = s.nextGreaterElement(nums1, nums2);
    for (int x : ans) {
        printf("%d\n", x);
    }
}

File: python/nextGreaterElement.py
from typing import List

'''
nums[::-1], 从右向左遍历
list.pop(), 移除元素，默认是最后一个元素
list.append(x), 从右侧加入
dict.get(x, -1), 如果不存在key返回-1
'''
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stk = []  # 单调递减栈
        d = {}  # value是key的next greater element
        for x in nums2[::-1]:
            print (x)
            while stk and stk[-1] < x:
                stk.pop()
            if stk:
                d[x] = stk[-1]
            stk.append(x)    
        return [d.get(x, -1) for x in nums1]
    
if __name__ == "__main__":
    nums1 = [4,1,2]
    nums2 = [1,3,4,2]
    print (Solution().nextGreaterElement(nums1, nums2))